Add FormData read & edit functions, write documentation

ghivert opened the pull request:
Hi!

- `FormData` was lacking some utilities function, and was lacking tests. That PR implements the remaining pieces in the most Gleamy possible way, while staying 100% compatible with JavaScript `FormData`, and implement tests for every `FormData` function.
- `gleam/fetch` was lacking documentation. That PR write documentation for every function, with examples.
- `README` was nice, but was lacking a bit of context, and could feel hard for newcomers. I took the opportunity of that PR to propose a new, improved version of the README. Of course it's subject to rewrite, modification or rejection if it does not suits you. 🙂 Feel free to be brutally honest to get the best README we can do!

---

About the README, I took freedom to write two parts "`gleam_fetch` for JS users" and "`gleam_fetch` for Erlang users", because `gleam_fetch` implies you have to think differently about your code (because of promises) if you're coming from Erlang, and it can be a bit different to what you're used to if you're coming from JS.
I tried to be friendly, to provide both an explanation and explaining the intentions behind the packages. I hesitated to write it elsewhere, but I think it's typically a place where it will be read, as HTTP requests are a fairly common requirements: having that information directly in the package would help users and not search for another source of information, like cookbook, etc. If you dislike it, or if you dislike this in this README, it can of course be removed or moved elsewhere.

ghivert commented on src/gleam/fetch/form_data.gleam:
````diff
@@ -1,20 +1,164 @@
+/// // -> False
+/// ```
+@external(javascript, "../../gleam_fetch_ffi.mjs", "hasFormData")
+pub fn has(form_data: FormData, key: String) -> Bool
````
Yes, well spotted! 💜

ghivert commented on src/gleam_fetch_ffi.mjs:
```diff
@@ -123,3 +125,39 @@ export function setBitsFormData(formData, key, value) {
+export async function getBitsFormData(formData, key) {
+  const data = [...formData.getAll(key)]
+  const blobs = data
+    .filter(value => value instanceof Blob)
```
That's an interesting point. I feel like it's not really how we want to use the package, but I could be wrong. In simplifile, you know what you're doing, and you read a file purposely as BitArray. Here, I'm a bit wondering if pushing a value as string, and getting it as BitArray is really consistent? I think unconsciously I would expect string values being read as string values, and vice-versa. 
Do you have any opinion on this? In any case, having the modification would be really straightforward.

ghivert replied:
Friendly reminder in case you forgot to answer. 💜

lpil replied:
I would not to not filter out that data, to remain consistent with the Gleam convention reading bits is a superset of reading text. e.g. simplifile and the existing text+binary functions in this package.

ghivert replied:
Thanks! Code has been modified to return everything from binaries.

ghivert commented:
> While the README text is great it's not in-keeping with the style of the other core packages' READMEs, and it teaches quite a lot of things that are not specifically this package, like error handling and concurrency. I think this content would be better somewhere else, but I'm not sure exactly where.

Is there any styleguide somewhere for the core packages' README?

If I understand correctly, you do prefer that I remove that content and keep a sleek README? I was thinking to build a documentation website for a while (for learning purpose, and for documenting my packages), maybe this could be the first step?

lpil commented:
There's no styleguide, but they concisely introduce the package and give any information specifically related to the package, but they do not teach Gleam more widely.

ghivert commented:
I just modified the README to get something more in-line with the current core packages, and I renamed `has` to `contains`. 
Let me know if something else is missing or wrong!

lpil commented on src/gleam/fetch/form_data.gleam:
````diff
@@ -1,20 +1,164 @@
+/// Remove a key and all its existing values.
+///
+/// ```gleam
+/// import gleam/fetch/form_data
````
```suggestion
```

lpil commented on CHANGELOG.md:
```diff
@@ -1,5 +1,10 @@
+## v1.2.0 - Unreleased
+
+- Add support for formdata reading and editing.
+- Add documentation for every functions and improve README.
```
This isn't a behaviour change 🙏 

```suggestion
```

lpil commented on README.md:
```diff
@@ -1,20 +1,32 @@
-import gleam/javascript/promise
+- Issue HTTP requests.
+- Handle HTTP responses in different formats (text, binary, JSON).
+- Read & Write [`FormData`](https://developer.mozilla.org/docs/Web/API/FormData).
```
Remove the features section please 🙏

lpil commented on README.md:
```diff
@@ -1,20 +1,32 @@
+> [!WARNING]
+> If you are running your Gleam project on the Erlang target (the default for
+> new Gleam projects) then you will want to use a different library which can
+> run on Erlang, such as [`gleam_httpc`](https://github.com/gleam-lang/httpc).
```
Move this to the bottom please 🙏

lpil commented on README.md:
````diff
@@ -32,3 +44,8 @@ pub fn main() {
 ```
+
+Documentation can be found at [https://hexdocs.pm/gleam_fetch](https://hexdocs.pm/gleam_fetch).
+
+`gleam_fetch` works on every JavaScript runtime implementing `fetch`, which
+implies all modern browsers, Node.js >= 18.0.0, Deno & Bun.
````
```suggestion
```

Fetch has been around for a long time so we don't need to warn people about support for it

lpil commented on src/gleam/fetch.gleam:
```diff
@@ -4,21 +4,61 @@ import gleam/http/request.{type Request}
+/// |> request.set_path("/example")
+/// |> fetch.to_fetch_request
+/// |> fetch.raw_send
+/// // -> Promise(Result(FetchResponse, FetchError))
```
```suggestion
```

There's no Promise constructor

lpil commented on src/gleam/fetch.gleam:
```diff
@@ -52,28 +124,140 @@ pub fn send_bits(
+///   |> request.set_host("example.com")
+///   |> request.set_path("/example")
+/// fetch.to_fetch_request(request)
+/// // -> FetchRequest
```
```suggestion
```

lpil commented on src/gleam/fetch.gleam:
```diff
@@ -52,28 +124,140 @@ pub fn send_bits(
+///     |> form_data.append("key", "value")
+///   })
+/// fetch.form_data_to_fetch_request(request)
+/// // -> FetchRequest
```
```suggestion
```

lpil commented on src/gleam/fetch.gleam:
```diff
@@ -52,28 +124,140 @@ pub fn send_bits(
+///   |> request.set_path("/example")
+///   |> request.set_body(<<"data">>)
+/// fetch.bitarray_request_to_fetch_request(request)
+/// // -> FetchRequest
```
```suggestion
```

lpil commented on src/gleam/fetch.gleam:
```diff
@@ -52,28 +124,140 @@ pub fn send_bits(
+/// |> fetch.to_fetch_request
+/// |> fetch.raw_send
+/// |> promise.map_try(fetch.from_fetch_response)
+/// // -> Promise(Result(Response(FetchBody), FetchError))
```
```suggestion
```

lpil commented on src/gleam/fetch/form_data.gleam:
````diff
@@ -1,20 +1,164 @@
+/// ensure correct conversions between JavaScript & Gleam.
 pub type FormData
 
+/// Create a `FormData` without any data.
+///
+/// ```gleam
+/// import gleam/fetch/form_data
+/// form_data.new()
+/// // -> form_data.FormData
+/// ```
````
```suggestion
/// Create a new empty `FormData`.
///
```

lpil commented on src/gleam/fetch/form_data.gleam:
````diff
@@ -1,20 +1,164 @@
+/// Append a key/string pair.
+///
+/// ```gleam
+/// import gleam/fetch/form_data
+/// form_data.new()
+/// |> form_data.append("key1", "value1")
+/// |> form_data.append("key1", "value2")
+/// |> form_data.append("key2", "value1")
+/// // "key1" -> ["value1", "value2"]
+/// // "key2" -> ["value1"]
````
```suggestion
/// form_data.new()
/// |> form_data.append("key1", "value1")
/// |> form_data.append("key1", "value2")
/// |> form_data.append("key2", "value1")
```

lpil commented on src/gleam/fetch/form_data.gleam:
````diff
@@ -1,20 +1,164 @@
+/// Append a key/bitarray pair.
+///
+/// ```gleam
+/// import gleam/fetch/form_data
+/// form_data.new()
+/// |> form_data.append_bits("key1", <<"value1">>)
+/// |> form_data.append_bits("key1", <<"value2">>)
+/// |> form_data.append_bits("key2", <<"value1">>)
+/// // "key1" -> [<<"value1">>, <<"value2">>]
+/// // "key2" -> [<<"value1">>]
````
```suggestion
/// form_data.new()
/// |> form_data.append_bits("key1", <<"value1">>)
/// |> form_data.append_bits("key1", <<"value2">>)
/// |> form_data.append_bits("key2", <<"value1">>)
```

lpil commented on src/gleam/fetch/form_data.gleam:
````diff
@@ -1,20 +1,164 @@
+/// Set key/string pair, and replace any existing value for the specified key.
+///
+/// ```gleam
+/// import gleam/fetch/form_data
+/// form_data.new()
+/// |> form_data.append("key1", "value1")
+/// |> form_data.append_bits("key1", <<"value2">>)
+/// |> form_data.set("key1", "value3")
+/// // "key1" -> ["value3"]
````
```suggestion
/// form_data.new()
/// |> form_data.append("key1", "value1")
/// |> form_data.append_bits("key1", <<"value2">>)
/// |> form_data.set("key1", "value3")
```

lpil commented on src/gleam/fetch/form_data.gleam:
````diff
@@ -1,20 +1,164 @@
+/// Set key/bitarray pair, and replace any existing value for the specified key.
+///
+/// ```gleam
+/// import gleam/fetch/form_data
+/// form_data.new()
+/// |> form_data.append("key1", "value1")
+/// |> form_data.append_bits("key1", <<"value2">>)
+/// |> form_data.set_bits("key1", <<"value3">>)
+/// // "key1" -> [<<"value3">>]
````
```suggestion
/// form_data.new()
/// |> form_data.append("key1", "value1")
/// |> form_data.append_bits("key1", <<"value2">>)
/// |> form_data.set_bits("key1", <<"value3">>)
```

lpil commented on src/gleam/fetch/form_data.gleam:
````diff
@@ -1,20 +1,164 @@
+/// binary values, take a look at [`get_bits`](#get_bits).
+///
+/// ```gleam
+/// import gleam/fetch/form_data
````
```suggestion
```

lpil commented on src/gleam/fetch/form_data.gleam:
````diff
@@ -1,20 +1,164 @@
+/// a `Promise`.
+///
+/// ```gleam
+/// import gleam/fetch/form_data
````
```suggestion
```

lpil commented on src/gleam_fetch_ffi.mjs:
```diff
@@ -123,3 +125,43 @@ export function setBitsFormData(formData, key, value) {
+    result.add(key)
+  }
+  return toList([...result])
+}
```
```suggestion
}
```

lpil commented on README.md:
```diff
@@ -1,20 +1,32 @@
+> new Gleam projects) then you will want to use a different library which can
+> run on Erlang, such as [`gleam_httpc`](https://github.com/gleam-lang/httpc).
+
+## Usage
```
```suggestion
```

ghivert commented:
Up to date!

lpil reviewed (approved):
Thank you! And sorry this took so long!

ghivert commented:
No problem! Thanks for your patience too! 💜
